### ISSUE TITLE
v2 changes

### ISSUE THREAD

**Issue by nicolaskruchten** [plotly/plotly.js#5395]
Here's what we're going to do for v2
- [ ] switch to `native-promise-only` and officially drop support for IE9 and IE10
- [ ] stop exporting D3
- [ ] stop exporting `Plotly.plot`
- [ ] drop `area`, `contourgl` and `pointcloud` traces
- [ ] drop support for polar plots in `scatter` (aka "micropolar"?) i.e. `scatter.r` and `scatter.t` and `layout.angularaxis` and `layout.radialaxis` and anything else related (cc @alexcjohnson anything else I missed?)

**Comment by nicolaskruchten** [comment]
@alexcjohnson suggested the following, which I approve if they take less than a day:

- [ ] change `isPlainObject` - either just drop the second clause or use something like the new one I found in https://github.com/plotly/plotly.js/issues/5151 - in principle this needn’t be a major but it feels a bit risky to me for some straight JS users
- [ ] drop some other exports: it also may be nice to strip `Plotly.Fx` to just `{hover, unhover, loneHover, loneUnhover}` and `Plotly.Plots` to just `{graphJson, resize}`

**Comment by nicolaskruchten** [comment]
Note to self: the plotly.com/javascript docs will need to stop relying on the built-in D3 :)

**Comment by nicolaskruchten** [comment]
@archmoj can you please provide a list as a comment in this issue of all the methods in our public API that have been dropped, so that folks like @cpsievert can easily grep for them in their own codebases. I've pinged the R/Julia/.Net downstream folks to let them know what's coming and directed them here :)

**Comment by archmoj** [comment]
> @archmoj can you please provide a list as a comment in this issue of all the methods in our public API that have been dropped, so that folks like @cpsievert can easily grep for them in their own codebases. I've pinged the R/Julia/.Net downstream folks to let them know what's coming and directed them here :)

Here is a list of `Plotly` methods that are going to be removed in v2:
```js
"Plotly.d3",

"Plotly.Plot",

"Plotly.Plots": [
    "modules",
    "allCategories",
    "allTypes",
    "subplotsRegistry",
    "transformsRegistry",
    "componentsRegistry",
    "layoutArrayContainers",
    "layoutArrayRegexes",
    "traceLayoutAttributes",
    "localeRegistry",
    "apiMethodRegistry",
    "collectableSubplotTypes",
    "register",
    "getModule",
    "traceIs",
    "getTransformIndices",
    "hasTransform",
    "getComponentMethod",
    "call",
    "attributes",
    "fontAttrs",
    "layoutAttributes",
    "fontWeight",
    "executeAPICommand",
    "computeAPICommandBindings",
    "manageCommandObserver",
    "hasSimpleAPICommandBindings",
    "redrawText",
    "previousPromises",
    "addLinks",
    "supplyDefaults",
    "supplyDefaultsUpdateCalc",
    "createTransitionData",
    "_hasPlotType",
    "cleanPlot",
    "linkSubplots",
    "clearExpandedTraceDefaultColors",
    "supplyDataDefaults",
    "supplyAnimationDefaults",
    "supplyAnimationFrameDefaults",
    "supplyAnimationTransitionDefaults",
    "supplyFrameDefaults",
    "supplyTraceDefaults",
    "hasMakesDataTransform",
    "supplyTransformDefaults",
    "supplyLayoutGlobalDefaults",
    "plotAutoSize",
    "supplyLayoutModuleDefaults",
    "purge",
    "style",
    "sanitizeMargins",
    "clearAutoMarginIds",
    "allowAutoMargin",
    "autoMargin",
    "doAutoMargin",
    "didMarginChange",
    "modifyFrames",
    "computeFrame",
    "recomputeFrameHash",
    "extendObjectWithContainers",
    "dataArrayContainers",
    "extendTrace",
    "extendLayout",
    "transition",
    "transitionFromReact",
    "doCalcdata",
    "rehover",
    "redrag",
    "generalUpdatePerTraceModule",
    "plotBasePlot",
    "cleanBasePlot"
],

"Plotly.Fx": [
    "moduleType",
    "name",
    "constants",
    "schema",
    "attributes",
    "layoutAttributes",
    "supplyLayoutGlobalDefaults",
    "supplyDefaults",
    "supplyLayoutDefaults",
    "calc",
    "getDistanceFunction",
    "getClosest",
    "inbox",
    "quadrature",
    "appendArrayPointValue",
    "castHoverOption",
    "castHoverinfo",
    "click"
],

"Plotly.Queue": [
    "add",
    "startSequence",
    "stopSequence",
    "undo",
    "redo",
    "plotDo"
],
```
Milestone: https://github.com/plotly/plotly.js/milestone/60?closed=1

**Comment by archmoj** [comment]
> Hmmm is `Plotly.purge()` still around? and `Plotly.extendTraces()` ? I would expect that the only function from this page https://plotly.com/javascript/plotlyjs-function-reference/ to be removed would be `Plotly.plot()` no? @alexcjohnson confirm your intentions here plz?

Yes `Plotly.purge()` and `Plotly.extendTraces() are still around.
The following methods are unchanged:
```js
"Plotly": [
    "version",
    "register",
    "newPlot",
    "restyle",
    "relayout",
    "redraw",
    "update",
    "react",
    "extendTraces",
    "prependTraces",
    "addTraces",
    "deleteTraces",
    "moveTraces",
    "purge",
    "addFrames",
    "deleteFrames",
    "animate",
    "setPlotConfig",
    "toImage",
    "validate",
    "downloadImage",
    "makeTemplate",
    "validateTemplate",
    "Icons",
    "Snapshot",
    "PlotSchema"
]
```

**Comment by nicolaskruchten** [comment]
Ah ok so in [your message above](https://github.com/plotly/plotly.js/issues/5395#issuecomment-763962569), it's `Plotly.Plots.purge` that is going away, and that's an alias to `Plotly.purge`?

Also, in the [message above](https://github.com/plotly/plotly.js/issues/5395#issuecomment-763962569), is `Plotly.Plot` the same thing as the `Plotly.plot()` function or something else? unsure if the capitalization matters here :)

**Comment by archmoj** [comment]
> Also, in the [message above](https://github.com/plotly/plotly.js/issues/5395#issuecomment-763962569), is `Plotly.Plot` the same thing as the `Plotly.plot()` function or something else? unsure if the capitalization matters here :)

Thanks for catching that! I edited the message above so that `Plotly.plot` is not capital.

**Comment by archmoj** [comment]
> Ah ok so in [your message above](https://github.com/plotly/plotly.js/issues/5395#issuecomment-763962569), it's `Plotly.Plots.purge` that is going away, and that's an alias to `Plotly.purge`?

No it is not an alias. 
`Plotly.purge()` performs other operations in addition to calling `Plotly.Plots.purge()`
https://github.com/plotly/plotly.js/blob/59390cbabf7d48c19179f68daf13e7d8af32f329/src/plot_api/plot_api.js#L3598-L3620

**Comment by nicolaskruchten** [comment]
Updates:

1. we're keeping `role: "object"` in the schema (because they're useful and consistent) and dropping the other `role` values (because they're inconsistent and therefore potentially dangerous to use and Plotly.py and docs don't use them)
2. we're keeping `pointcloud` and marking it deprecated
3. we're marking `heatmapgl` as deprecated

**Comment by dhirschfeld** [comment]
Just wondering if https://github.com/plotly/plotly.js/pull/5230 will make it in 2.0?

**Comment by archmoj** [comment]
> Just wondering if #5230 will make it in 2.0?

I think new features like that could possibly be added in a minor afterwards e.g. v2.x.0
